pkg/storage/storagewrappers/sharediterator/shared_iterator_datastore.go

```diff
@@ -0,0 +1,337 @@
+
+	sf.internalStorage.mu.Lock()
+
+	keyItem, found := sf.internalStorage.iters[cacheKey]
+
+	if found {
+		keyItem.counter++
+		sf.internalStorage.mu.Unlock()
+		span.SetAttributes(attribute.Bool("found", true))
+
+		// For now, this will be identical to the not found path.  When the SF iterator datastore
+		// is fully implemented, the found path will look up the shared iterator and create
+		// a clone from it.
+	} else {
+		if len(sf.internalStorage.iters) >= sf.internalStorage.limit {
+			sf.internalStorage.mu.Unlock()
+			sharedIteratorBypassed.WithLabelValues(storagewrappersutil.OperationReadStartingWithUser).Inc()
+			// we cannot share this iterator because we have reached the size limit.
+			return sf.RelationshipTupleReader.ReadStartingWithUser(ctx, store, filter, options)
+		}
+
+		sf.internalStorage.iters[cacheKey] = &internalSharedIterator{
+			counter: 1,
+		}
+		sf.internalStorage.mu.Unlock()
+		span.SetAttributes(attribute.Bool("found", false))
+	}
+	actual, err := sf.RelationshipTupleReader.ReadStartingWithUser(ctx, store, filter, options)
+	if err != nil {
+		go func() {
+			_, _ = sf.deref(cacheKey)
+		}()
+		return nil, err
+	}
+	sharedIteratorQueryHistogram.WithLabelValues(
+		storagewrappersutil.OperationReadStartingWithUser, strconv.FormatBool(found),
+	).Observe(float64(time.Since(start).Milliseconds()))
+
+	return newSharedIterator(sf, cacheKey, actual), nil
+}
+
+func (sf *IteratorDatastore) ReadUsersetTuples(
+	ctx context.Context,
+	store string,
+	filter storage.ReadUsersetTuplesFilter,
+	options storage.ReadUsersetTuplesOptions,
+) (storage.TupleIterator, error) {
+	ctx, span := tracer.Start(
+		ctx,
+		"sharedIterator.ReadUsersetTuples",
+	)
+	defer span.End()
+	start := time.Now()
+
+	cacheKey := storagewrappersutil.ReadUsersetTuplesKey(store, filter)
+	span.SetAttributes(attribute.String("cache_key", cacheKey))
+
+	sf.internalStorage.mu.Lock()
+
+	keyItem, found := sf.internalStorage.iters[cacheKey]
+
+	if found {
+		keyItem.counter++
+		sf.internalStorage.mu.Unlock()
+		span.SetAttributes(attribute.Bool("found", true))
+
+		// For now, this will be identical to the not found path.  When the SF iterator datastore
+		// is fully implemented, the found path will look up the shared iterator and create
+		// a clone from it.
+	} else {
+		if len(sf.internalStorage.iters) >= sf.internalStorage.limit {
+			sf.internalStorage.mu.Unlock()
+			sharedIteratorBypassed.WithLabelValues(storagewrappersutil.OperationReadUsersetTuples).Inc()
+			// we cannot share this iterator because we have reached the size limit.
+			return sf.RelationshipTupleReader.ReadUsersetTuples(ctx, store, filter, options)
+		}
+
+		sf.internalStorage.iters[cacheKey] = &internalSharedIterator{
+			counter: 1,
+		}
+		sf.internalStorage.mu.Unlock()
+		span.SetAttributes(attribute.Bool("found", false))
+	}
+	actual, err := sf.RelationshipTupleReader.ReadUsersetTuples(ctx, store, filter, options)
+	if err != nil {
+		go func() {
```

Comment:
Why does this need to launch a goroutine for something that is relatively quick?

Reply:
removed go func